### PR TITLE
[Graph] New incoming edge storage + sanity checks + component interning

### PR DESCRIPTION
**Pull request by leogrim**
@stkem @yingjieMiao

### CODE REVIEW

**Comment by stkem** [comment]
discussed in person. lgtm, perhaps with a comment or two

File: kifi-backend/modules/graph/app/com/keepit/graph/simple/BufferedMap.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.keepit.graph.simple
+
+import scala.collection.mutable.{ Map => MutableMap, Set => MutableSet }
+
+class BufferedMap[A, B](current: MutableMap[A, B], updated: MutableMap[A, B] = MutableMap[A, B](), removed: MutableSet[A] = MutableSet[A]()) extends MutableMap[A, B] {
+  def get(key: A) = if (removed.contains(key)) None else updated.get(key) orElse current.get(key)
+  def iterator = updated.iterator ++ current.iterator.withFilter { case (key, _) => !hasBuffered(key) }
```
Comment (yingjieMiao) [inline kifi-backend/modules/graph/app/com/keepit/graph/simple/BufferedMap.scala:7]: unrelated to this pull request. the `withFilter` part is not executed until you start using the iterator. i.e. the `hasBuffered()` has undefined behavior until you start using the iterator.

File: kifi-backend/modules/graph/app/com/keepit/graph/simple/MutableVertex.scala [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    new MutableOutgoingEdges(mutableEdges)
+  }
+
+  def apply(outgoingEdges: OutgoingEdges): MutableOutgoingEdges = {
+    val mutableEdges = MutableMap[Component, MutableMap[VertexId, EdgeDataReader]]()
+    outgoingEdges.edges.foreach {
+      case (component, destinations) =>
+        val mutableDestinations = MutableMap[VertexId, EdgeDataReader]() ++= destinations
+        mutableEdges += (component -> mutableDestinations)
+    }
```
Comment (yingjieMiao) [inline kifi-backend/modules/graph/app/com/keepit/graph/simple/MutableVertex.scala:108]: Maybe it's clearer to encapsulate these low level map-key-value operations into some graph-level api, e.g. `addDestinationsToComponent`, `getDestinationsByComponent`, etc. Looks like these map operations are repeated in many places.

File: kifi-backend/modules/graph/app/com/keepit/graph/simple/BufferedMap.scala (hunk 1 above)
Comment (yingjieMiao) [inline kifi-backend/modules/graph/app/com/keepit/graph/simple/BufferedMap.scala:7]: as discussed in person, the `hasBuffered()` behavior changes while we iterate, and having graph writes in the same time. Looks like it's OK for current usage, but you may want to add some caution notes there ;-)

**Comment by yingjieMiao** [comment]
lgtm!

**Pull request merged**